kyo-prelude/shared/src/main/scala/kyo2/Result.scala

```diff
@@ -4,134 +4,159 @@ import Result.*
-                            Failure(ex)
+                            Panic(ex)
 
-        inline def flatten[B](using ev: A <:< Result[B]): Result[B] =
+        def flatten[E2, U](using ev: T <:< Result[E2, U]): Result[E | E2, U] =
             flatMap(ev)
 
-        inline def map[B](inline f: A => B): Result[B] =
+        inline def map[U](inline f: T => U): Result[E, U] =
             flatMap(v => Result.success(f(v)))
 
-        inline def filter(inline p: A => Boolean): Result[A] =
+        inline def withFilter(inline p: T => Boolean): Result[E | NoSuchElementException, T] =
+            filter(p)
+
+        inline def filter(inline p: T => Boolean): Result[E | NoSuchElementException, T] =
             flatMap { v =>
                 if !p(v) then
-                    throw new NoSuchElementException("Predicate does not hold for " + v)
-                v
+                    Panic(new NoSuchElementException("Predicate does not hold for " + v))
+                else
+                    v
             }
 
-        inline def recover[B >: A](inline rescueException: PartialFunction[Throwable, B]): Result[B] =
-            (self: @unchecked) match
-                case Failure(exception) if rescueException.isDefinedAt(exception) =>
-                    Result(rescueException(exception))
-                case _ => self.asInstanceOf[Result[B]]
-
-        inline def recoverWith[B >: A](inline rescueException: PartialFunction[Throwable, Result[B]]): Result[B] =
-            (self: @unchecked) match
-                case Failure(exception) if rescueException.isDefinedAt(exception) =>
-                    rescueException(exception)
-                case _ => self.asInstanceOf[Result[B]]
-
-        inline def toEither: Either[Throwable, A] =
-            fold(Left(_))(Right(_))
-
-        inline def toTry: Try[A] =
-            fold(scala.util.Failure(_))(scala.util.Success(_))
+        inline def recover[U >: T](inline pf: PartialFunction[Cause[E], U]): Result[E, U] =
+            try
+                (self: @unchecked) match
+                    case self: Cause[E] if pf.isDefinedAt(self) =>
+                        Result.success(pf(self))
+                    case _ => self
+            catch
+                case ex: Throwable if NonFatal(ex) =>
+                    Panic(ex)
+
+        inline def recoverWith[E2, U >: T](inline pf: PartialFunction[Cause[E], Result[E2, U]]): Result[E | E2, U] =
+            try
+                (self: @unchecked) match
+                    case self: Cause[E] if pf.isDefinedAt(self) =>
+                        pf(self)
+                    case _ => self
+            catch
+                case ex: Throwable if NonFatal(ex) =>
+                    Panic(ex)
+
+        def toEither: Either[E | Throwable, T] =
+            fold(Left(_))(Left(_))(Right(_))
+
+        def toTry: Try[T] =
+            fold(e => scala.util.Failure(new NoSuchElementException(s"Cause: $e")))(scala.util.Failure(_))(scala.util.Success(_))
```

Comment:
I think we should use Either's implementation for inspiration and require evidence of `E <: Throwable`